### DISCUSSION TITLE
About release versions

### DISCUSSION THREAD

**Discussion by weyderfs** [tgenv/tgenv#7] [Ideas]
Guys,
I'm thinking about when starts versioning the solution. What're you think about that?

Actually the `tgenv` is working at branch `main`.

**Comment by iokiwi** [comment]
I'm struggling to think of real need for versioning right now. Maybe if we start pushing it to homebrew it could be necessary for the package manager to be aware of the package version but imo for now main is okay?

Unless you want a `stable` branch to use for installs by default.

So develop against main and merge/cherrypick to stable for release?

**Reply by weyderfs** [reply]
> I'm struggling to think of real need for versioning right now. Maybe if we start pushing it to homebrew it could be necessary for the package manager to be aware of the package version but imo for now main is okay?
> 
> Unless you want a `stable` branch to use for installs by default.
> 
> So develop against main and merge/cherrypick to stable for release?

@iokiwi thank you for share your opinion, so let's continue in this way for now without versioning.

**Reply by mdinos** [reply]
If you're trying to install this in a CI environment, it'd be good to be able to just pull a particular tag/branch. That way, you don't have to pull the whole commit history, speeding up your build. 

Example:

```
git clone --depth 1 --branch v0.0.1 https://github.com/tgenv/tgenv.git ~/.tgenv
```

I know the history isn't very big at the moment, so it's probably not a huge problem though.

At the moment, to pin the version you'd have to do something like this:

```
git clone https://github.com/tgenv/tgenv.git ~/.tgenv && pushd ~/.tgenv && git reset --hard ${COMMIT_SHA} && popd
```

**Reply by iokiwi** [reply]
Hey @mdinos. Thanks for sharing your ideas.

We could add `--depth=1` to the install command in the README so users download less by default. We should check if this conflicts with the instructions for upgrade command using git pull.

Readme would read something like this.

```bash
git clone -depth=1 --branch=main https://github.com/tgenv/tgenv.git ~/.tgenv
```

There is merit to tagging commits and allowing users pin a specific version rather than always download whatever is at the HEAD of main branch.

I want to emphasise though that this should just be to enable hermetic installs for our users if they desire but we should only commit to supporting the HEAD of main lest we create a large maintenance burden for ourselves.

This is a bit of a speil not specifically directed at you, @mdinos.

I don't see that that policy should be a problem in the immediate or long term as - beyond bugfixes, maintenance and expanding runtime support - I don't invision this project needs to be nor should be very fast moving with many breaking changes. `tgenv` is a simple tool to do a single task well and I would loathe to see it become a bloated mess.

We should be kind and show love to our future selves and our successors and KISS (keep it simple, stupid).

Thats just my 2c.

**Reply by iokiwi** [reply]
In terms of tagging what do you think?

Semver_ish_

 * `major.minor.bugfix`? `breaking.major.minor`? How do we define major or minor?
 * We could start with 0.x.y as an indication of the immaturity of this project
 * We could start from 1.x.y as an indicator of a basic level of maturity
 
Timestamp?
 
  * ISO-8601 E.g 2023-01-26
  * Super simple, just a label, save a lot of time worrying about definitions of which version number to increment.
  * Can tag automatically on merge to main (I think this is possible)

**Comment by iokiwi** [comment]
On the other hand, semverish seems to work for tfenv. Proposing we tag ``v1.0.0`` asap so we can start pinning versions (I too have this requirement at work)

**Reply by weyderfs** [reply]
@iokiwi to me okay starts with `v1.0.0`, when you quote about _I too have this requirement at work_, what's mean?

**Reply by iokiwi** [reply]
as @mdinos described its useful to be able to pin a tagged version rather than just suck down the latest thing from github. 

better for stability and consistency, better for security.

**Reply by iokiwi** [reply]
We have our first tag. v1.0.0. Sorry I thought I did this ages ago, @mdinos 

I've raised an PR here to update the readme with instructions to pull a tagged release. https://github.com/tgenv/tgenv/pull/25
